fusio/src/impls/disk/tokio/fs.rs

```diff
@@ -1,20 +1,106 @@
-use std::io;
+use std::{
+    ffi::OsString,
+    io,
+    io::ErrorKind,
+    path::{Path as StdPath, PathBuf},
+    pin::Pin,
+    sync::atomic::{AtomicU64, Ordering},
+};
 
 use async_stream::stream;
+use fusio_core::MaybeSendFuture;
 use futures_core::Stream;
 use tokio::{
-    fs::{create_dir_all, remove_file},
+    fs::{create_dir_all, remove_file, File, OpenOptions as TokioOpenOptions},
     task::spawn_blocking,
 };
 
 use crate::{
     disk::tokio::TokioFile,
     durability::DirSync,
     error::Error,
-    fs::{FileMeta, FileSystemTag, Fs, OpenOptions},
+    fs::{CasCondition, FileMeta, FileSystemTag, Fs, FsCas, OpenOptions},
     path::{path_to_local, Path},
 };
 
+static TMP_COUNTER: AtomicU64 = AtomicU64::new(0);
+
+fn metadata_tag(meta: &std::fs::Metadata) -> String {
+    use std::time::{SystemTime, UNIX_EPOCH};
+
+    let len = meta.len();
+    let modified = meta.modified().unwrap_or(SystemTime::UNIX_EPOCH);
+    let duration = modified.duration_since(UNIX_EPOCH).unwrap_or_default();
+    format!(
+        "{:016x}:{:016x}:{:08x}",
+        len,
+        duration.as_secs(),
+        duration.subsec_nanos()
+    )
+}
+
+async fn current_tag(path: &StdPath) -> Result<String, Error> {
+    let meta = tokio::fs::metadata(path).await?;
+    Ok(metadata_tag(&meta))
+}
+
+fn with_suffix(path: &StdPath, suffix: &str) -> PathBuf {
+    let mut os: OsString = path.as_os_str().to_os_string();
+    os.push(suffix);
+    PathBuf::from(os)
+}
+
+async fn write_atomic(path: &StdPath, payload: &[u8]) -> Result<(), Error> {
+    let parent = path.parent().ok_or_else(|| {
+        Error::Path(Box::new(io::Error::new(
+            ErrorKind::InvalidInput,
+            "path has no parent",
+        )))
+    })?;
+
+    let unique = TMP_COUNTER.fetch_add(1, Ordering::Relaxed);
+    let tmp_name = format!(".tmp-{}-{}", std::process::id(), unique);
+    let tmp_path = parent.join(tmp_name);
+
+    tokio::fs::write(&tmp_path, payload).await?;
+
+    if tokio::fs::try_exists(path).await? {
+        tokio::fs::remove_file(path).await?;
```

Comment:
do not need to remove file at first, "remove then rename" is not atomic, there is a middle sate without any file, simply rename it would be fine.